docs/proposals/qos-management/orm-nri/20240303-orm-nri.md

```diff
@@ -0,0 +1,368 @@
+
+In this part, the method based on the Kubelet API polling is referred to as 
+**_Bypass_** Mode, while another method based on NRI is referred to as **_NRI_** Mode.
+
+#### Addon
+
+- The ORM support two operational modes: Bypass or NRI. Only one mode can be active 
+at any given time. When creating a new ORM Manger, the current operational mode can 
+be determined by reading the configuration, and it does not support changing the 
+mode during runtime.
+    
+    ```go
+    const AGENT_MODE_NRI = "nri"
+    const AGENT_NODE_BYPASS = "bypass"
+    
+    type ManagerImpl struct {
+    	ctx context.Context
+      ....
+      // ORM run mode: bypass or nri.
+      // Bypass mode is triggered by polling kubelet api to get the pod event.
+      // NRI mode is required containerd version >= 1.7.0 and NRI enabled.
+    	mode string
+      ....
+    }
+    
+    func NewManger(... config *config.Configuration){
+    
+    	if config.GenericConfiguration.NRIEnable && ValidateNRIStatus(){
+    		m.mode=AGENT_MODE_NRI
+    	}else{
+    		m.mode=AGENT_NODE_BYPASS
+    	}
+    }
+    
+    func ValidateNRIStatus() bool{
+    // TODO detect NRI status
+    }
+    ```
+    
+- The ORM ManagerImpl functions as an NRI stub, implementing processing logic 
+within the corresponding hook event functions.
+    
+    ```go
+    import "github.com/containerd/nri/pkg/stub"
+    
+    type ManagerImpl struct {
+        ctx context.Context
+        ....
+        // nriStub is the implementtion of NRI events handlers
+        nriStub stub.Stub
+        // nriMask stores the specific events that need to be hooked
+        nriMask stub.EventMask
+        ....
+    }
+    ```
+    
+- In enhancing the ORM implementation, three hook functions are required: 
+`RunPodSandbox()`, `CreateContainer()`, and `RemovePodSandbox()`.
+    
+  **Step 1**, during `RunPodSanbox()`, the `Admit()` function is triggered. 
+If `Admit()` succeeds, resources are allocated for the container, and the pod 
+creation process continues. If `Admit()` fails, pod creation also fails.
+    ```go
+        func (m *MangerImpl) RunPodSandbox(podSandbox *api.PodSandbox) error {
+    	    // get podspec from podSandbox -> MetaServer?
+                // Admit
+    	    return m.topologyManager.Admit(pod)
+        }
+    ```
+
+  **Step 2**, after a successful `Admit()`, the process proceeds to the 
+`CreateContainer()` event. At this point, resources have been allocated for the 
+container by `Admit()`. The corresponding resources are updated in the container's 
+spec and returned.
+  ```go
+      func (m *MangerImpl) CreateContainer(pod *api.PodSandbox, container *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+        // Update Container Spec from the podResoures
+        adjust, err:= m.updateContainer(pod, container)
+  	    return adjust, nil, err
+      }
+  ```
+
+  **Step 3**, During `RemovePodSandbox()`, all resource allocations related to 
```

Comment:
What is the handling mechanism if the `RemovePodSandbox` call fails? We may need to try to ensure that `RemovePodSandbox` is called successfully. If the call fails, the resources allocated to the container may not be released, and the states recorded in the QRM plugin and SysAdvisor may be incorrect.

Reply:
The reconcile mechanism will handle this problem.